src/Core/Checkout/Cart/Order/RecalculationService.php

```diff
@@ -88,7 +91,11 @@ public function recalculateOrder(string $orderId, Context $context, array $sales
         }
 
         // change scope to be able to write protected state fields of transactions and deliveries
-        $context->scope(Context::SYSTEM_SCOPE, function (Context $context) use ($orderData): void {
+        $context->scope(Context::SYSTEM_SCOPE, function (Context $context) use ($orderData, $order): void {
+            if (($lineItems = $order->getLineItems()) instanceof OrderLineItemCollection) {
+                $this->orderLineItemRepository->delete(array_values($lineItems->fmap(static fn (OrderLineItemEntity $lineItem) => ['id' => $lineItem->getId()])), $context);
```

Comment:
Can you maybe add a testcase for this one? I think it is quite critical and would help understand the problem.

And also i would suggest to not directly delete all lineitems from the original order, but rather perform a check which lineitems exist in the original order, that are not part of the recalculated order and delete only those, this makes it also more explicit why we need to the deletion

Reply:
also esp for huge orders this approach should have a performance benefit

Reply:
Hi @keulinho i believe you are right? I adjusted the code and created a testcase